# Practical 3
# Course Code and name: 2CS501 and Machine Learning
# Name and Roll No. : Dev Bachani 21BCE049


# AIM: Simple and Multiple Linear Regression using Gradient Descent & Normal Equation Method (without using sklearn or equivalent library for both)


In [76]:
import numpy as np
import random
import pandas as pd
import matplotlib.pylab as plt
from sklearn import datasets, metrics
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_absolute_error,mean_squared_error

In [77]:
data=pd.read_csv("BostonHousing.csv")
data

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio       b  lstat  medv  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

In [78]:
rows=data.shape[0]
columns=data.shape[1]
training_examples=int(0.8*rows)
test_examples=rows-training_examples
print(rows,training_examples,test_examples)

506 404 102


In [79]:
X_train=data.iloc[0:training_examples,0:columns-1] #Dataframe
Y_train=data.iloc[0:training_examples,columns-1:] #Dataframe
# X_train=data.iloc[0:training_examples,0] #Series
# Y_train=data.iloc[0:training_examples,1] #Series


X_test=data.iloc[training_examples:,0:columns-1]
Y_test=data.iloc[training_examples:,columns-1:]


# X_train_np=np.array(X_train.iloc[:,0]) # to be used with X_train=data.iloc[0:training_examples,0:1]
X_train_np=np.array(X_train.iloc[:]) # to be used with X_train=data.iloc[0:training_examples,0]
Y_train_np=np.array(Y_train.iloc[:])

In [80]:
X_train
X_train_np=np.ones(shape=(X_train.shape[0],X_train.shape[1]+1))
X_train_np[:,1:]=X_train
X_train_np,X_train.shape

(array([[1.00000e+00, 6.32000e-03, 1.80000e+01, ..., 1.53000e+01,
         3.96900e+02, 4.98000e+00],
        [1.00000e+00, 2.73100e-02, 0.00000e+00, ..., 1.78000e+01,
         3.96900e+02, 9.14000e+00],
        [1.00000e+00, 2.72900e-02, 0.00000e+00, ..., 1.78000e+01,
         3.92830e+02, 4.03000e+00],
        ...,
        [1.00000e+00, 1.42362e+01, 0.00000e+00, ..., 2.02000e+01,
         3.96900e+02, 2.03200e+01],
        [1.00000e+00, 9.59571e+00, 0.00000e+00, ..., 2.02000e+01,
         3.76110e+02, 2.03100e+01],
        [1.00000e+00, 2.48017e+01, 0.00000e+00, ..., 2.02000e+01,
         3.96900e+02, 1.97700e+01]]),
 (404, 13))

In [81]:
Y_train
Y_train_np=np.array(Y_train)
Y_train_np,Y_train_np.shape

(array([[24. ],
        [21.6],
        [34.7],
        [33.4],
        [36.2],
        [28.7],
        [22.9],
        [27.1],
        [16.5],
        [18.9],
        [15. ],
        [18.9],
        [21.7],
        [20.4],
        [18.2],
        [19.9],
        [23.1],
        [17.5],
        [20.2],
        [18.2],
        [13.6],
        [19.6],
        [15.2],
        [14.5],
        [15.6],
        [13.9],
        [16.6],
        [14.8],
        [18.4],
        [21. ],
        [12.7],
        [14.5],
        [13.2],
        [13.1],
        [13.5],
        [18.9],
        [20. ],
        [21. ],
        [24.7],
        [30.8],
        [34.9],
        [26.6],
        [25.3],
        [24.7],
        [21.2],
        [19.3],
        [20. ],
        [16.6],
        [14.4],
        [19.4],
        [19.7],
        [20.5],
        [25. ],
        [23.4],
        [18.9],
        [35.4],
        [24.7],
        [31.6],
        [23.3],
        [19.6],
        [18.7],
        [16. ],
        

In [82]:
X_test
X_test_np=np.ones(shape=(X_test.shape[0],X_test.shape[1]+1))
X_test_np[:,1:]=X_test
X_test_np,X_test_np.shape


(array([[1.00000e+00, 4.15292e+01, 0.00000e+00, ..., 2.02000e+01,
         3.29460e+02, 2.73800e+01],
        [1.00000e+00, 6.79208e+01, 0.00000e+00, ..., 2.02000e+01,
         3.84970e+02, 2.29800e+01],
        [1.00000e+00, 2.07162e+01, 0.00000e+00, ..., 2.02000e+01,
         3.70220e+02, 2.33400e+01],
        ...,
        [1.00000e+00, 6.07600e-02, 0.00000e+00, ..., 2.10000e+01,
         3.96900e+02, 5.64000e+00],
        [1.00000e+00, 1.09590e-01, 0.00000e+00, ..., 2.10000e+01,
         3.93450e+02, 6.48000e+00],
        [1.00000e+00, 4.74100e-02, 0.00000e+00, ..., 2.10000e+01,
         3.96900e+02, 7.88000e+00]]),
 (102, 14))

In [83]:
Y_test
Y_test_np=np.array(Y_test)
Y_test_np,Y_test_np.shape

(array([[ 8.5],
        [ 5. ],
        [11.9],
        [27.9],
        [17.2],
        [27.5],
        [15. ],
        [17.2],
        [17.9],
        [16.3],
        [ 7. ],
        [ 7.2],
        [ 7.5],
        [10.4],
        [ 8.8],
        [ 8.4],
        [16.7],
        [14.2],
        [20.8],
        [13.4],
        [11.7],
        [ 8.3],
        [10.2],
        [10.9],
        [11. ],
        [ 9.5],
        [14.5],
        [14.1],
        [16.1],
        [14.3],
        [11.7],
        [13.4],
        [ 9.6],
        [ 8.7],
        [ 8.4],
        [12.8],
        [10.5],
        [17.1],
        [18.4],
        [15.4],
        [10.8],
        [11.8],
        [14.9],
        [12.6],
        [14.1],
        [13. ],
        [13.4],
        [15.2],
        [16.1],
        [17.8],
        [14.9],
        [14.1],
        [12.7],
        [13.5],
        [14.9],
        [20. ],
        [16.4],
        [17.7],
        [19.5],
        [20.2],
        [21.4],
        [19.9],
        

In [84]:
rows_train=X_train_np.shape[0]
columns_train=X_train_np.shape[1]
rows_train,columns_train

(404, 14)

In [619]:
# theta_Gradient_Descent=np.random.uniform(0, 1, size=columns_train)
# theta_Gradient_Descent


array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [660]:
iterations=20000
alpha=0.000001
update=np.zeros(X_train_np.shape[1])

theta_Gradient_Descent=np.zeros(columns_train)
# theta_Gradient_Descent=np.array([ 1.32375124e+00, -2.17177941e-01,  4.58088924e-02,  2.36054141e-02,
#         1.60040470e+00,  8.05806984e-02,  6.07388137e+00, -5.84302758e-03,
#        -9.35175700e-01,  3.57643092e-01, -1.37325559e-02, -4.51144388e-01,
#         1.03168036e-02, -4.74233152e-01])

Y_train_np_reshaped=Y_train_np.reshape(-1)
for i in range(iterations):
    pred = np.dot(X_train_np, theta_Gradient_Descent)
    
    
    
    # print(Y_train_np_reshaped.shape)

    error = pred - Y_train_np_reshaped

    for j in range(columns_train):
        gradient = np.dot(error, X_train_np[:, j])
        theta_Gradient_Descent[j] = theta_Gradient_Descent[j] - (alpha / rows_train) * gradient

        
    
        
pred = np.dot(X_train_np, theta_Gradient_Descent)
print(mean_absolute_error(Y_train_np_reshaped, pred))
print(mean_squared_error(Y_train_np_reshaped,pred))



5.327165958997935
53.65985301405008


In [658]:
theta_Gradient_Descent

array([ 1.71178143e+00, -2.15837636e-01,  4.58112210e-02,  2.19433034e-02,
        1.75487898e+00, -2.17298103e-02,  6.04331745e+00, -5.91627060e-03,
       -9.39493424e-01,  3.56786934e-01, -1.37145408e-02, -4.53956056e-01,
        1.01725532e-02, -4.75275557e-01])

<h5>
array([ 1.71178143e+00, -2.15837636e-01,  4.58112210e-02,  2.19433034e-02,
        1.75487898e+00, -2.17298103e-02,  6.04331745e+00, -5.91627060e-03,
       -9.39493424e-01,  3.56786934e-01, -1.37145408e-02, -4.53956056e-01,
        1.01725532e-02, -4.75275557e-01])
</h5>

In [631]:
# Normal Equation
# Theta = (Xt . X )^ -1 * Xt * y
X_train_np_T=X_train_np.T
Xt_X=np.dot(X_train_np_T,X_train_np)
# print(Xt_X.shape)

try:
  Xt_X_inverse= np.linalg.inv(Xt_X)
except:
  Xt_X_inverse=np.linalg.pinv(Xt_X)

Xt_Y=np.dot(X_train_np_T,Y_train_np)
# print(Xt_Y.shape)

theta_Normal_Eqn=np.dot(Xt_X_inverse,Xt_Y)
# print(Xt_Y)

prediction_Normal_Eqn=np.dot(X_test_np,theta_Normal_Eqn)

print(mean_absolute_error(Y_test_np,prediction_Normal_Eqn))
print(mean_squared_error(Y_test_np,prediction_Normal_Eqn))

4.730017250960253
32.79986268019678


In [652]:
theta_Normal_Eqn

array([[ 3.00771669e+01],
       [-2.02135297e-01],
       [ 4.41276341e-02],
       [ 5.26739364e-02],
       [ 1.88474315e+00],
       [-1.49281487e+01],
       [ 4.76038673e+00],
       [ 2.88734527e-03],
       [-1.30025278e+00],
       [ 4.61661953e-01],
       [-1.55434673e-02],
       [-8.11632369e-01],
       [-1.97174433e-03],
       [-5.32273431e-01]])

In [656]:
theta_Normal_Eqn.reshape(-1)

array([ 3.00771669e+01, -2.02135297e-01,  4.41276341e-02,  5.26739364e-02,
        1.88474315e+00, -1.49281487e+01,  4.76038673e+00,  2.88734527e-03,
       -1.30025278e+00,  4.61661953e-01, -1.55434673e-02, -8.11632369e-01,
       -1.97174433e-03, -5.32273431e-01])

# Results
<h1>Using Gradient Descent:</h1>
<h3>Theata=[ 1.32375124e+00, -2.17177941e-01,  4.58088924e-02,  2.36054141e-02,
        1.60040470e+00,  8.05806984e-02,  6.07388137e+00, -5.84302758e-03,
       -9.35175700e-01,  3.57643092e-01, -1.37325559e-02, -4.51144388e-01,
        1.03168036e-02, -4.74233152e-01]</h3>

<br>
<h4> iterations=2000000</h4>
<h4> mean_absolute_error=3.2309158774330418</h4>
<h4> mean_squared_error=24.0290007435647</h4>

<br>

<h1>Using Normal Equation:</h1>
<h3>Theata=[ 3.00771669e+01, -2.02135297e-01,  4.41276341e-02,  5.26739364e-02,
        1.88474315e+00, -1.49281487e+01,  4.76038673e+00,  2.88734527e-03,
       -1.30025278e+00,  4.61661953e-01, -1.55434673e-02, -8.11632369e-01,
       -1.97174433e-03, -5.32273431e-01]</h3>

<h4> mean_absolute_error=4.730017250960253</h4>
<h4> mean_squared_error=32.79986268019678</h4>
